File: packages/python/identities/starfish_identities/trusted_adders.py
"""Trusted-adder allow-list helpers.

When opening a keyring, the caller must supply the set of Ed25519 pubkeys whose
``addedBy`` signatures are trusted for CEK-wrap entries. This module derives that
list from the ownership relationship between a device and the space owner.
"""

from __future__ import annotations


def compute_owner_trusted_adders(
    owner_ed_pub: str | None,
    self_ed_pub: str,
) -> list[str]:
    """Return the trusted-adder allow-list for a device opening an owned space's keyring.

    For the root device (or any device where ``owner_ed_pub`` equals ``self_ed_pub``),
    only ``self_ed_pub`` is trusted — no external adder needs to be in the list.

    For a paired/linked device, the space was created by the root (``owner_ed_pub``)
    and the linked device's own key (``self_ed_pub``) was added later.  Both must be
    trusted so the device can read both the original root-added entries and any
    entries added after it joined.

    Args:
        owner_ed_pub: Ed25519 pubkey (hex) of the space owner (root). When ``None``,
            defaults to ``self_ed_pub`` (treats device as its own owner).
        self_ed_pub: Ed25519 pubkey (hex) of this device.

    Returns:
        ``[self_ed_pub]`` when ``owner_ed_pub`` equals (or defaults to) ``self_ed_pub``,
        otherwise ``[owner_ed_pub, self_ed_pub]``.
    """
    owner = owner_ed_pub if owner_ed_pub is not None else self_ed_pub
    return [owner, self_ed_pub] if owner != self_ed_pub else [self_ed_pub]


__all__ = ["compute_owner_trusted_adders"]

File: packages/ts/identities/src/trusted-adders.ts
/**
 * Owner trusted-adder allow-list for opening/sealing an OWNED keyring.
 *
 * When the session's owner key equals the device key (the common single-device
 * case) the allow-list is just `[selfEdPub]`; otherwise the owner key is also
 * trusted so that a paired device can open keyrings sealed by the root identity.
 *
 * Kept in its own leaf module so both `client.ts`-style helpers and session
 * builders in consuming packages can import it without risking circular deps.
 */

/**
 * Compute the trusted-adder allow-list for the caller's OWNED keyring.
 *
 * @param ownerEdPub - The Ed25519 pubkey of the owner (root identity).
 *   Undefined is treated as "same as self" (i.e., self is the owner).
 * @param selfEdPub - The Ed25519 pubkey of the current device.
 */
export function computeOwnerTrustedAdders(ownerEdPub: string | undefined, selfEdPub: string): string[] {
  const owner = ownerEdPub ?? selfEdPub
  return owner !== selfEdPub ? [owner, selfEdPub] : [selfEdPub]
}
